sbi/utils/sbiutils.py

```diff
@@ -105,92 +107,83 @@ def match_shapes_of_theta_and_x(
+        num_samples: Desired number of samples.
+        show_progressbar: Whether to show a progressbar during sampling.
+        warn_acceptance: A minimum acceptance rate under which to warn about slowness.
 
     Returns:
-        Accepted samples, and estimated acceptance
-         probability
+        Accepted samples and acceptance rate as scalar Tensor.
     """
 
-    assert (
-        not posterior_nn.training
-    ), "posterior nn is in training mode, but has to be in eval mode for sampling."
+    assert not posterior_nn.training, "Posterior nn must be in eval mode for sampling."
 
-    samples = []
-    num_remaining = num_samples
-    num_sampled_total = 0
-    leakage_warning_raised = False
-
-    # we might not always want a progressbar for sampling. E.g. we want to display that,
-    # after each round, we also draw samples just for logging. Thus, the progressbar
-    # can be turned off with the show_progressbar argument.
+    # Progress bar can be skipped, e.g. when sampling after each round just for logging.
     pbar = tqdm(
-        total=num_remaining,
         disable=not show_progressbar,
-        desc="Drawing {0} posterior samples".format(num_remaining),
+        total=num_samples,
+        desc=f"Drawing {num_samples} posterior samples",
     )
 
-    with pbar:
-        while num_remaining > 0:
-
-            # XXX: we need this reshape here because posterior_nn.sample sometimes return
-            # leading singleton dimension instead of (num_samples), e.g., (1, 10000, 4)
-            # instead of (10000, 4). and this can't be handled by MultipleIndependent. issue #141
-            sample = posterior_nn.sample(num_remaining, context=x).reshape(
-                num_remaining, -1
+    num_remaining, num_sampled_total = num_samples, 0
+    accepted, acceptance_rate = [], float("Nan")
+    leakage_warning_raised = False
+    # In each iteration of the loop we sample the remaining number of samples from the
+    # posterior. Some of these samples have 0 probability under the prior, i.e. there
+    # is leakage (acceptance rate<1) so sample again until reaching `num_samples`.
+    while num_remaining > 0:
+
+        candidates = posterior_nn.sample(num_remaining, context=x)
+        # TODO we need this reshape here because posterior_nn.sample sometimes return
+        # leading singleton dimension instead of (num_samples), e.g., (1, 10000, 4)
+        # instead of (10000, 4). This can't be handled by MultipleIndependent, see #141.
+        candidates = candidates.reshape(num_remaining, -1)
+        num_sampled_total += num_remaining
+
+        are_within_prior = torch.isfinite(prior.log_prob(candidates))
+        accepted.append(candidates[are_within_prior])
+
+        num_accepted = are_within_prior.sum().item()
+        pbar.update(num_accepted)
+
+        # To avoid endless sampling when leakage is high, we raise a warning if the
+        # acceptance rate is too after the first 1.000 samples.
```

Comment:
1000 or 1_000 or 1,000 but please not 1.000 ;)

Also missing a 'high' after "is too"

Reply:
Got it!

Reply:
Wait, too _low_!

Reply:
Ah, yes, of course